tests/containers/kubernetes_utils.py

```diff
@@ -0,0 +1,467 @@
+from kubernetes.dynamic import DynamicClient, ResourceField
+
+import ocp_resources.pod
+import ocp_resources.deployment
+import ocp_resources.service
+import ocp_resources.persistent_volume_claim
+import ocp_resources.project_request
+import ocp_resources.namespace
+import ocp_resources.project_project_openshift_io
+import ocp_resources.deployment
+import ocp_resources.resource
+import ocp_resources.pod
+import ocp_resources.namespace
+import ocp_resources.project_project_openshift_io
+import ocp_resources.project_request
+
+from tests.containers import socket_proxy
+
+
+class TestFrameConstants:
+    GLOBAL_POLL_INTERVAL_MEDIUM = 10
+    TIMEOUT_2MIN = 2 * 60
+
+
+logging.basicConfig(level=logging.DEBUG)
+LOGGER = logging.getLogger(__name__)
+
+
+# https://github.com/RedHatQE/openshift-python-wrapper/tree/main/examples
+
+def get_client() -> kubernetes.dynamic.DynamicClient:
+    try:
+        # client = kubernetes.dynamic.DynamicClient(client=kubernetes.config.new_client_from_config())
+        # probably same as above
+        client = ocp_resources.resource.get_client()
+        return client
+    except kubernetes.config.ConfigException as e:
+        # probably bad config
+        logging.error(e)
+    except kubernetes.dynamic.exceptions.UnauthorizedError as e:
+        # wrong or expired credentials
+        logging.error(e)
+    except kubernetes.client.ApiException as e:
+        # unexpected, we catch unauthorized above
+        logging.error(e)
+    except Exception as e:
+        # unexpected error, assert here
+        logging.error(e)
+
+    raise RuntimeError("Failed to instantiate client")
+
+
+def get_username(client: kubernetes.dynamic.DynamicClient) -> str:
+    # can't just access
+    # > client.configuration.username
+    # because we normally auth using tokens, not username and password
+
+    # this is what kubectl does (see kubectl -v8 auth whoami)
+    self_subject_review_resource: kubernetes.dynamic.Resource = client.resources.get(
+        api_version="authentication.k8s.io/v1", kind="SelfSubjectReview"
+    )
+    self_subject_review: kubernetes.dynamic.ResourceInstance = client.create(self_subject_review_resource)
+    username: str = self_subject_review.status.userInfo.username
+    return username
+
+
+class TestKubernetesUtils:
+    def test_get_username(self):
+        client = get_client()
+        username = get_username(client)
+        assert username is not None and len(username) > 0
+
+
+class TestFrame:
+    def __init__[T](self):
+        self.stack: list[tuple[T, Callable[[T], None] | None]] = []
+
+    def defer_resource[T: ocp_resources.resource.Resource](self, resource: T, wait=False,
+                                                           destructor: Callable[[T], None] | None = None) -> T:
+        result = resource.deploy(wait=wait)
+        self.defer(resource, destructor)
+        return result
+
+    def add[T](self, resource: T, destructor: Callable[[T], None] = None) -> T:
```

Comment:
Is this used somewhere? Also, would it be better to name this something like `add_defer` or similar? :thinking:

Reply:
I still haven't decided whether I like this approach in general.

I do dislike it less than what opendatahub-tests is doing with the fixtuers...

Reply:
removed add method